### PR TITLE
EZP-31079: Provided validation for email uniqueness and login pattern

### PR DESCRIPTION
**Pull request by Nattfarinn**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31079](https://jira.ez.no/browse/EZP-31079)
| **Bug/Improvement**| yes
| **New feature**    | yes
| **Target version** | `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

### Requires:
 - https://github.com/ezsystems/ezplatform-admin-ui/pull/1288
 - https://github.com/ezsystems/ezplatform-content-forms/pull/17
 ### Supplemented by:
 - https://github.com/ezsystems/ezplatform-user/pull/62

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/FieldType/User/Type.php [hunk 1]
```diff
@@ -353,6 +375,32 @@ public function validate(FieldDefinition $fieldDefinition, SPIValue $fieldValue)
...
+                $email = $fieldValue->email;
+                try {
+                    $user = $this->userHandler->loadByEmail($email);
+                } catch (LogicException $exception) {
+                    // There are multiple users with the same email
+                }
+
+                // Don't prevent email update
+                if ($user->id != $fieldValue->contentId) {
+                    // If you want to change this ValidationError message, please remember to change it also in Repository Forms in lib/Validator/Constraints/FieldValueValidatorMessages class
```
[the remark this answers is not in the document]
Comment (Nattfarinn) [reply eZ/Publish/Core/FieldType/User/Type.php]: True. Haven't updated this comment. Change is reflected on Content Forms already.

File: eZ/Publish/API/Repository/Tests/UserServiceTest.php [hunk 2]
```diff
@@ -1084,6 +1079,118 @@ public function testCreateUserThrowsInvalidArgumentException()
...
+        $editorsGroupId = $this->generateId('group', 13);
+        /* BEGIN: Use Case */
+        // $editorsGroupId is the ID of the "Editors" user group in an eZ
+        // Publish demo installation
+
+        $userService = $repository->getUserService();
+
+        // Instantiate a create struct with mandatory properties
+        $userCreate = $userService->newUserCreateStruct(
+        // admin is an existing login
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/UserServiceTest.php]: Nitpick: Wrong indent

```suggestion
            // admin is an existing login
```

File: eZ/Publish/API/Repository/Tests/UserServiceTest.php [hunk 3]
```diff
@@ -1084,6 +1079,118 @@ public function testCreateUserThrowsInvalidArgumentException()
...
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/UserServiceTest.php]:
```suggestion
            // admin is an existing login
```

File: eZ/Publish/API/Repository/Tests/UserServiceTest.php [hunk 4]
```diff
@@ -1084,6 +1079,118 @@ public function testCreateUserThrowsInvalidArgumentException()
...
+        $this->fail('Expected ValidationError messages did not occur.');
+    }
+
+    /**
+     * Test for the createUser() method.
+     *
+     * @covers \eZ\Publish\API\Repository\UserService::createUser
+     * @depends eZ\Publish\API\Repository\Tests\UserServiceTest::testCreateUser
+     */
+    public function testCreateInvalidFormatUsername()
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/UserServiceTest.php]:
```suggestion
    public function testCreateInvalidFormatUsername(): void
```

File: eZ/Publish/API/Repository/Tests/UserServiceTest.php [hunk 5]
```diff
@@ -1084,6 +1079,118 @@ public function testCreateUserThrowsInvalidArgumentException()
         $this->fail('Expected ValidationError messages did not occur.');
     }
 
+    /**
+     * Test for the createUser() method.
+     *
+     * @covers \eZ\Publish\API\Repository\UserService::createUser
+     * @depends eZ\Publish\API\Repository\Tests\UserServiceTest::testCreateUser
+     */
+    public function testCreateUserWithEmailAlreadyTaken()
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/UserServiceTest.php]:
```suggestion
    public function testCreateUserWithEmailAlreadyTaken(): voidd
```

File: eZ/Publish/API/Repository/Tests/UserServiceTest.php [hunk 6]
```diff
@@ -1084,6 +1079,118 @@ public function testCreateUserThrowsInvalidArgumentException()
...
+        ]);
+
+        $existingUser = $this->createUserVersion1(
+            'existing_user',
+            'unique@email.com',
+            $userContentType,
+        );
+
+        $editorsGroupId = $this->generateId('group', 13);
+        /* BEGIN: Use Case */
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/UserServiceTest.php]: ` /* BEGIN: Use Case */` and `/* END: Use Case */` are obsolete.

**Comment by Nattfarinn** [comment]
Closing in favour of: https://github.com/ezsystems/ezplatform-kernel/pull/3

**Pull request closed without merging**